libr/bin/format/wasm/wasm.c

```diff
@@ -105,6 +105,60 @@ static bool consume_str_new(RBuffer *b, ut64 bound, ut32 *len_out, char **str_ou
 	return false;
 }
 
+/*
+ * Wasm Names are utf-8 character strings. This means '7\x00' is a valid name 2
+ * byte name. R2 uses tcc to do C like function declarations, so we encode
+ * these functions in a way that can be decoded 1 to 1. Encoding is for char
+ * '\xXX' that is not allowed, we encode it as "_XX_" Where XX is [0-9A-Z] (no
+ * lower). Should the original function have a substring of the form "_XX_" we
+ * encode it as "_5F_XX_".
+ */
+#define WASM_IS_DIGIT(c) (c >= '0' && c <= '9')
+#define WASM_IS_ENC_HEX(c) (WASM_IS_DIGIT (c) || (c >= 'A' && c <= 'F'))
+#define WASM_AFTER_U_NOT_HEX(str, i) (!WASM_IS_ENC_HEX (str[i + 1]) || !WASM_IS_ENC_HEX (str[i + 2]))
+#define WASM_AFTER_UNDERSCORE_OK(str, i) (WASM_AFTER_U_NOT_HEX (str, i) || str[i + 3] != '_')
+#define WASM_IS_ALPH(c) ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z'))
+#define WASM_IS_OK_EXCEPT__(loc, c) (WASM_IS_ALPH (c) || (loc != 0 && WASM_IS_DIGIT (c))) // C functions can't start with [0-9]
+#define WASM_UNDERSCORE_OK(str, i, max) (str[i] == '_' && (max - i < 4 || WASM_AFTER_UNDERSCORE_OK (str, i)))
+#define WASM_IS_OK(str, i, max) (WASM_IS_OK_EXCEPT__ (i, str[i]) || WASM_UNDERSCORE_OK (str, i, max))
+static bool consume_encoded_name_new(RBuffer *b, ut64 bound, ut32 *len_out, char **str_out) {
+	ut32 len;
+	char *orig = NULL;
+	if (!consume_str_new (b, bound, &len, &orig)) {
+		return false;
+	}
+
+	// room for even every character getting encoded
+	char *sout = malloc (len * 4 + 2);
+	if (!sout) {
+		free (orig);
+		return false;
+	}
+
+	size_t i, oi = 0;
+	for (i = 0; i < len; i++) {
+		if (WASM_IS_OK (orig, i, len)) {
+			sout[oi++] = orig[i];
+		} else {
+			oi += sprintf (sout + oi, "_%02x_", orig[i]);
```

Comment:
use snprintf passingn len - oi as size, maybe check for overflow may be good in the other case

Reply:
Should be fixed, along with a pointer mistake

Reply:
Another fix... I should use length of sout rather then length of orig.